base/mpfr.jl

```diff
@@ -237,58 +237,113 @@ function _duplicate(x::BigFloat)
     return z
 end
 
+function is_bigfloat_special(x::Real)
+    # x is special if it is a zero, an Inf, or a NaN
+    return iszero(x) || !isfinite(x)
+end
+
+function BigFloat_with_special_values(x::Real, precision::Integer=DEFAULT_PRECISION[])
+    z = BigFloat(; precision=precision)
+    z.sign = 1-2*signbit(x)
+
+    if iszero(x)
+        z.exp = mpfr_special_exponent_zero
+    elseif isinf(x)
+        z.exp = mpfr_special_exponent_inf
+    else
+        z.exp = mpfr_special_exponent_nan
+    end
+
+    return z
+end
+
+function store_significand_in_limbs!(z::BigFloat, significand::UInt64)
+    nlimbs = (z.prec + 8*Core.sizeof(Limb) - 1) ÷ (8*Core.sizeof(Limb))
+
+    # Limb is a CLong which is a UInt32 on Windows which makes this more complicated and slower.
+    if Limb === UInt64
+        for i in 1:nlimbs-1
+            unsafe_store!(z.d, 0x0, i)
+        end
+        unsafe_store!(z.d, significand, nlimbs)
+    else
+        for i in 1:nlimbs-2
+            unsafe_store!(z.d, 0x0, i)
+        end
+        unsafe_store!(z.d, significand % UInt32, nlimbs-1)
+        unsafe_store!(z.d, (significand >> 32) % UInt32, nlimbs)
+    end
+end
+
 # convert to BigFloat
 for (fJ, fC) in ((:si,:Clong), (:ui,:Culong))
     @eval begin
-        function BigFloat(x::($fC), r::MPFRRoundingMode=ROUNDING_MODE[]; precision::Integer=DEFAULT_PRECISION[])
+        function BigFloat_mpfr_fallback(x::($fC), r::MPFRRoundingMode=ROUNDING_MODE[]; precision::Integer=DEFAULT_PRECISION[])
             z = BigFloat(;precision=precision)
             ccall(($(string(:mpfr_set_,fJ)), libmpfr), Int32, (Ref{BigFloat}, $fC, MPFRRoundingMode), z, x, r)
             return z
         end
     end
 end
 
-function BigFloat(x::Float64, r::MPFRRoundingMode=ROUNDING_MODE[]; precision::Integer=DEFAULT_PRECISION[])
-    z = BigFloat(;precision)
+function BigFloat(x::Base.BitInteger64, r::MPFRRoundingMode=ROUNDING_MODE[]; precision::Integer=DEFAULT_PRECISION[])
+    if is_bigfloat_special(x)
+        return BigFloat_with_special_values(x, precision)::BigFloat
+    end
+
+    unsigned_x = unsigned(abs(x))
```

Comment:
rather than manipulating values manually, can we just use `decompose` here and then it should work for arbitrary number types.

Reply:
Not sure if I quite follow. `decompose(x::Integer)` returns `(x, 0, 1)`.